fpga_interchange/lookahead.cc

```diff
@@ -0,0 +1,1522 @@
+        gzFile file = gzopen(filename.c_str(), "r");
+        if (file == Z_NULL) {
+            return false;
+        }
+
+        std::vector<uint8_t> output_data;
+        output_data.resize(4096);
+        std::stringstream sstream(std::ios_base::in | std::ios_base::out | std::ios_base::binary);
+        while (true) {
+            int ret = gzread(file, output_data.data(), output_data.size());
+            NPNR_ASSERT(ret >= 0);
+            if (ret > 0) {
+                sstream.write((const char *)output_data.data(), ret);
+                NPNR_ASSERT(sstream);
+            } else {
+                NPNR_ASSERT(ret == 0);
+                int error;
+                gzerror(file, &error);
+                NPNR_ASSERT(error == Z_OK);
+                break;
+            }
+        }
+
+        NPNR_ASSERT(gzclose(file) == Z_OK);
+
+        sstream.seekg(0);
+        kj::std::StdInputStream istream(sstream);
+        capnp::InputStreamMessageReader message_reader(istream, reader_options);
+
+        lookahead_storage::Lookahead::Reader lookahead = message_reader.getRoot<lookahead_storage::Lookahead>();
+        return from_reader(chipdb_hash, lookahead);
+    } else {
+        boost::iostreams::mapped_file_source file;
+        try {
+            file.open(filename.c_str());
+        } catch (std::ios_base::failure &fail) {
+            return false;
+        }
+
+        if (!file.is_open()) {
+            return false;
+        }
+
+        const char *data = reinterpret_cast<const char *>(file.data());
+        const kj::ArrayPtr<const ::capnp::word> words =
+                kj::arrayPtr(reinterpret_cast<const ::capnp::word *>(data), file.size() / sizeof(::capnp::word));
+        ::capnp::FlatArrayMessageReader reader(words, reader_options);
+        lookahead_storage::Lookahead::Reader lookahead = reader.getRoot<lookahead_storage::Lookahead>();
+        return from_reader(chipdb_hash, lookahead);
+    }
+}
+
+void Lookahead::write_lookahead(const std::string &chipdb_hash, const std::string &file) const
+{
+    ::capnp::MallocMessageBuilder message;
+
+    lookahead_storage::Lookahead::Builder lookahead = message.initRoot<lookahead_storage::Lookahead>();
+    to_builder(chipdb_hash, lookahead);
+    write_message(message, file);
+}
+
+void Lookahead::init(const Context *ctx, DeterministicRNG *rng)
+{
+    std::string lookahead_filename;
+    if (kUseGzipForLookahead) {
+        lookahead_filename = ctx->args.chipdb + ".lookahead.tgz";
+    } else {
+        lookahead_filename = ctx->args.chipdb + ".lookahead";
+    }
+
+    std::string chipdb_hash = ctx->get_chipdb_hash();
+
+    if (ctx->args.rebuild_lookahead || !read_lookahead(chipdb_hash, lookahead_filename)) {
+        build_lookahead(ctx, rng);
+        if (!ctx->args.dont_write_lookahead) {
+            write_lookahead(chipdb_hash, lookahead_filename);
+        }
+    }
+}
+
+bool safe_add_i32(int32_t a, int32_t b, int32_t *out) { return !__builtin_add_overflow(a, b, out); }
```

Comment:
this will need a MSVC/other compiler fallback

Reply:
Done.  I took a look at boost's safe int option, and it is basically undocumented, so not a good option here.